package packethandler;

// define types of messages
enum MessageType {
  SHARE=0;
  ADD_CONTACT_RQST=1;
  GENERAL=2;
  ADD_CONTACT_RESPONSE=3;
  INSTANT_MSG=4;
  INSTANT_FILE=5;
  DELETE_CONTACT_NOTIF=6;
}

message GenericPacket {
  required string data = 1;
  required string signature = 2;
}

message BufferPacketInfo {
  required string owner = 1;
  required string ownerPublicKey = 2;
  repeated string users = 3;
  optional bool online = 4;
}

message BufferPacketMessage {
  required string sender_id = 1;
  optional string sender_public_key = 2;
  required string RSAenc_key = 3;
  required string AESenc_message = 4;
  required MessageType type = 5;
}

message BufferPacket {
  repeated GenericPacket owner_info = 1;
  repeated GenericPacket messages = 2;
}

message ValidatedBufferPacketMessage {
  required string sender = 1;
  required string message = 2;
  required string index = 3;
  required MessageType type = 4;
}

message ContactInfo {
  optional string name = 1;
  optional string birthday = 2;
  optional string office_number = 3;
  optional string gender = 4;
  optional int32 country = 5;
  optional string city = 6;
  optional int32 language = 7;
}

message InstantMessage {
  required string sender = 1;
  required string message = 2;
  required int32 date = 3;
  required MessageType type = 4;
}

message InstantFileMessage {
  required string ser_mdm = 1;
  required string ser_dm = 2;
  required string filename = 3;
  optional string sender_msg = 4;
}

message StoreMessagesResult {
  required string result = 1;
  required int32 stored_msgs = 2;
  repeated string failed = 3;
}

message CreateMSIDResult {
  required string result = 1;
  optional string name = 2;
  optional string private_key = 3;
  optional string public_key = 4;
}

message PrivateShareNotification {
  required string name = 1;
  required string msid = 2;
  required string public_key = 3;
  optional string private_key = 4;
  repeated string admins = 5;
  repeated string readonlys = 6;
  required string dir_db_key = 7;
}
